src/MagnumPlugins/WebPImporter/WebPImporter.cpp

```diff
@@ -0,0 +1,168 @@
+#include <webp/mux_types.h>
+
+#include <Corrade/Containers/Optional.h>
+#include <Corrade/Containers/ScopeGuard.h>
+#include <Corrade/Utility/Algorithms.h>
+#include <Corrade/Utility/Debug.h>
+#include <Magnum/PixelFormat.h>
+#include <Magnum/Math/Functions.h>
+#include <Magnum/Trade/ImageData.h>
+
+namespace Magnum { namespace Trade {
+
+WebPImporter::WebPImporter() = default;
+
+WebPImporter::WebPImporter(PluginManager::AbstractManager& manager, const Containers::StringView& plugin): AbstractImporter{manager, plugin} {}
+
+WebPImporter::~WebPImporter() = default;
+
+ImporterFeatures WebPImporter::doFeatures() const { return ImporterFeature::OpenData; }
+
+bool WebPImporter::doIsOpened() const { return _in; }
+
+void WebPImporter::doClose() { _in = nullptr; }
+
+void WebPImporter::doOpenData(Containers::Array<char>&& data, DataFlags dataFlags) {
+    /* Because here we're copying the data and using the _in to check if file
+       is opened, having them nullptr would mean openData() would fail without
+       any error message. It's not possible to do this check on the importer
+       side, because empty file is valid in some formats (OBJ or glTF). We also
+       can't do the full import here because then doImage2D() would need to
+       copy the imported data instead anyway (and the uncompressed size is much
+       larger). This way it'll also work nicely with a future openMemory(). */
+    if(data.isEmpty()) {
+        Error{} << "Trade::WebpImporter::openData(): the file is empty";
+        return;
+    }
+
+    /* Take over the existing array or copy the data if we can't */
+    if(dataFlags & (DataFlag::Owned|DataFlag::ExternallyOwned)) {
+        _in = std::move(data);
+    } else {
+        _in = Containers::Array<char>{NoInit, data.size()};
+        Utility::copy(data, _in);
+    }
+}
+
+UnsignedInt WebPImporter::doImage2DCount() const { return 1; }
+
+Containers::Optional<ImageData2D> WebPImporter::doImage2D(UnsignedInt, UnsignedInt) {
+
+    /* Create the structure that contains the WebP image data. */
+    WebPData webp_data;
+    webp_data.bytes = reinterpret_cast<uint8_t*>(_in.data());
+    webp_data.size = _in.size();
+
+    /* Verify the file is a WebP image file */
+    bool isWebPImage = WebPGetInfo(webp_data.bytes, webp_data.size, nullptr, nullptr);
+    if(!isWebPImage) {
+        Error() << "Trade::WebpImporter::image2D(): wrong file format";
+        return Containers::NullOpt;
+    }
+
+    /* Reading the file information */
+    //WebPBitstreamFeatures info;
+    //CORRADE_ASSERT(WebPGetFeatures(webp_data.bytes, webp_data.size, &info) == VP8_STATUS_OK, "The features couldn't be retrieved.", Containers::NullOpt);
+
+    WebPDecoderConfig config;
+    CORRADE_ASSERT(WebPInitDecoderConfig(&config), "Decoder version mismatch", Containers::NullOpt);
+    CORRADE_ASSERT(WebPGetFeatures(webp_data.bytes, webp_data.size, &config.input) == VP8_STATUS_OK, "The features couldn't be retrieved.", Containers::NullOpt);
```

Comment:
There's an important distinction between assertions, internal assertions and runtime errors. 

Assertions are meant to check either programmer errors (say, passing a negative value to `sqrt()`) in which case the `CORRADE_ASSERT()` macros with a descriptive message is used, or assumptions about program state that are not expected to fail, which is then checked with the `CORRADE_INTERNAL_ASSERT()` macro. As the internal assertion is not expected to fire for a library user, the macro has no message argument (in contrast it prints a file/line info which is useful for the library author). And finally, there's `*_ASSERT_OUTPUT()` macros, which don't get compiled away in builds that have assertions disabled, which is useful if you're wrapping an actual function call and not just a comparison.

Then, runtime errors are meant to check external conditions that are out of programmer control -- files not found, files being corrupted, etc. Compared to assertions, these don't abort the program, jumping into a debugger, but instead fail gracefully, printing an error message and returning a failure.

So, the above should be instead like this -- the first one isn't expected to fail as it doesn't depend on the file data (if I see correctly), and the case of a decoder version mismatch is so rare and hard to test that I don't think we should bother at this point :) Because the call needs to be done even with assertions disabled, it needs to be an `*_OUTPUT()` macro. The second one, as far as I can see, may fail if the file data is corrupted, so that should be a runtime error instead:

```cpp
    CORRADE_INTERNAL_ASSERT_OUTPUT(WebPInitDecoderConfig(&config));
    
    if(WebPGetFeatures(webp_data.bytes, webp_data.size, &config.input) != VP8_STATUS_OK) {
        Error{} << "The features couldn't be retrieved.";
        return {};
    }
```

The third assertion macro below needs the same treatment -- also a runtime error.

And a tip -- you can use `return {}` as a shorthand for error return. It'll be `false`, `NullOpt` or `nullptr` depending on what the return type is, all signaling a failure.